synapse/streams/events.py

```diff
@@ -21,20 +23,43 @@
 from synapse.handlers.typing import TypingNotificationEventSource
 from synapse.types import StreamToken
 
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
 
-class EventSources:
-    SOURCE_TYPES = {
-        "room": RoomEventSource,
-        "presence": PresenceEventSource,
-        "typing": TypingNotificationEventSource,
-        "receipt": ReceiptEventSource,
-        "account_data": AccountDataEventSource,
-    }
 
-    def __init__(self, hs):
-        self.sources: Dict[str, Any] = {
-            name: cls(hs) for name, cls in EventSources.SOURCE_TYPES.items()
-        }
+@attr.s(frozen=True, slots=True, auto_attribs=True)
+class _EventSourcesInner:
+    room: RoomEventSource
+    presence: PresenceEventSource
+    typing: TypingNotificationEventSource
+    receipt: ReceiptEventSource
+    account_data: AccountDataEventSource
+
+    def get_sources(
+        self,
+    ) -> Generator[
+        Tuple[
+            str,
+            Union[
+                RoomEventSource,
+                PresenceEventSource,
+                TypingNotificationEventSource,
+                ReceiptEventSource,
+                AccountDataEventSource,
+            ],
+        ],
+        None,
+        None,
+    ]:
```

Comment:
I think this below is valid, and a little bit less intimidating (especially since we're not making use of being able to pass values into the generator, or returning from it):

```suggestion
    ) -> Iterator[
        Tuple[
            str,
            Union[
                RoomEventSource,
                PresenceEventSource,
                TypingNotificationEventSource,
                ReceiptEventSource,
                AccountDataEventSource,
            ],
        ],
    ]:
```

Reply:
Yes, I recently learned you can use `Iterator` here.